examples/control-vector-generator/control-vector-generator.cpp

```diff
@@ -0,0 +1,763 @@
+#include "common.h"
+#include "llama.h"
+#include "ggml.h"
+
+#include <cstdio>
+#include <string>
+#include <tuple>
+#include <vector>
+#include <algorithm>
+#include <iostream>
+#include <fstream>
+
+// TODO read everything over and make sure it makes sense because you're dropping logic errors left and right
+
+struct diff_wrapper {
+    float * diff;   // matrix of size [n_rows, cb_data.n_embd] with zero rows stripped
+    size_t n_rows;  // number of rows in the matrix for size calculation
+};
+
+/* TODO part of multithreading
+struct tokens_pair {
+    size_t max_seq_len;
+    std::string positive;
+    std::string negative;
+    std::vector<llama_token> tokens_pos;
+    std::vector<llama_token> tokens_neg;
+}; */
+
+struct callback_data {
+    std::vector<uint8_t> data;
+
+    int n_tokens = 0;
+    int n_embd = 0;
+    bool is_eval_pos = true;
+
+    // each element of the vector correspond to one layer
+    std::vector<float *> v_pos;       // vector of matrices of size [n_embd, n_tokens]
+    std::vector<float *> v_neg;       // vector of matrices of size [n_embd, n_tokens]
+    std::vector<float *> v_final;     // vector of finished vectors of size [n_embd]
+    std::vector<diff_wrapper> v_diff; // vector of matrices of size [n_embd, m] where m ~ n_tokens * n_completions
```

Comment:
Just thinking, I initially use `std::vector<float *>` for more visibility. But now seeing that it doing more harm than good, should we change all of these to `ggml_tensor` right now? (I know that we eventually need to change them to tensor, just don't know when)

In either way, congrats and thanks for your work @christianazinn . We already have this to somewhat working, now just need to improve code & performance. It's unfortunately that I don't have much time to dive deep into this. But feel free to let me know if you need help.

Reply:
I can start on changing the code to use GGML rather than stdlib this weekend, so I can change them now. To start I will just need to know what the most basic useful functions are for get/set, operations (in particular I would like to know how exactly mul_mat works vs. mul_mat_inplace because inplace vector multiplication would be much faster), etc. Thanks for putting up with my lack of knowledge.